// @r35007/mock-server/dist/middlewares/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelperMiddlewares = exports.Rewriter = exports.PageNotFound = exports.Initializer = exports.ErrorHandler = exports.Defaults = void 0;
const defaults_1 = __importDefault(require("./defaults"));
exports.Defaults = defaults_1.default;
const errorHandler_1 = __importDefault(require("./errorHandler"));
exports.ErrorHandler = errorHandler_1.default;
const HelperMiddlewares = __importStar(require("./HelperMiddlewares"));
exports.HelperMiddlewares = HelperMiddlewares;
const initializer_1 = __importDefault(require("./initializer"));
exports.Initializer = initializer_1.default;
const pageNotFound_1 = __importDefault(require("./pageNotFound"));
exports.PageNotFound = pageNotFound_1.default;
const rewriter_1 = __importDefault(require("./rewriter"));
exports.Rewriter = rewriter_1.default;
