// @r35007/mock-server/dist/defaults.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Store = exports.Rewriters = exports.Injectors = exports.Middlewares = exports.Db = exports.Config = void 0;
const path = __importStar(require("path"));
const middlewares_1 = require("./middlewares");
exports.Config = {
    base: '',
    bodyParser: true,
    cookieParser: true,
    dbMode: 'mock',
    homePage: true,
    host: 'localhost',
    id: 'id',
    log: false,
    logger: true,
    noCache: true,
    noCors: false,
    noGzip: false,
    port: 3000,
    quiet: false,
    readOnly: false,
    reverse: false,
    root: process.cwd(),
    static: path.join(process.cwd(), 'public'), // Path to host a static files
};
exports.Db = {};
exports.Middlewares = middlewares_1.HelperMiddlewares;
exports.Injectors = [];
exports.Rewriters = {};
exports.Store = {};
